node/src/block_producer/block_producer_reducer.rs

```diff
@@ -519,35 +294,424 @@ impl BlockProducerEnabled {
+                        gt_pred_sub_window && lt_next_sub_window
+                    } else {
+                        gt_pred_sub_window || lt_next_sub_window
+                    };
+                    if is_same_global_sub_window || are_windows_overlapping && !within_range {
+                        density
+                    } else {
+                        0
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            let grace_period_end = grace_period_end(pred_block.constants());
+            let min_window_density = if is_same_global_sub_window
+                || curr_global_slot_since_hard_fork.slot_number.as_u32() < grace_period_end
+            {
+                pred_consensus_state.min_window_density
+            } else {
+                let cur_density = current_sub_window_densities.iter().sum();
+                let min_density = pred_consensus_state
+                    .min_window_density
+                    .as_u32()
+                    .min(cur_density);
+                UnsignedExtendedUInt32StableV1(min_density.into())
+            };
+
+            let next_sub_window_densities = current_sub_window_densities
+                .into_iter()
+                .enumerate()
+                .map(|(i, density)| (i as u32, density))
+                .map(|(i, density)| {
+                    let is_next_sub_window = i == next_relative_sub_window;
+                    if is_next_sub_window {
+                        let density = if is_same_global_sub_window {
+                            density
+                        } else {
+                            0
+                        };
+                        if incr_window {
+                            density + 1
+                        } else {
+                            density
+                        }
+                    } else {
+                        density
+                    }
+                })
+                .map(|v| UnsignedExtendedUInt32StableV1(v.into()))
+                .collect();
+
+            (min_window_density, next_sub_window_densities)
+        };
+
+        let consensus_state = ConsensusProofOfStakeDataConsensusStateValueStableV2 {
+            blockchain_length: UnsignedExtendedUInt32StableV1((pred_block.height() + 1).into()),
+            epoch_count,
+            min_window_density,
+            sub_window_densities,
+            last_vrf_output: vrf_truncated_output,
+            total_currency: (&total_currency).into(),
+            curr_global_slot_since_hard_fork,
+            global_slot_since_genesis,
+            staking_epoch_data,
+            next_epoch_data,
+            has_ancestor_in_same_checkpoint_window,
+            block_stake_winner,
+            block_creator,
+            coinbase_receiver,
+            // TODO(binier): Staged_ledger.can_apply_supercharged_coinbase_exn
+            supercharge_coinbase: constraint_constants().supercharged_coinbase_factor != 0,
```

Comment:
hmmm out of scope for this PR but this logic is not right. And we already have the above function implemented already (but cannot be called from here)